File: js/packages/mechanisms/casper/test/exact/server/scheme.test.ts
import { describe, expect, it } from "vitest";
import { ExactCasperScheme } from "../../../src/exact/server/scheme";

const testAsset = "aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788";
const testPayTo = "00aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788";
const testNetwork = "casper:casper-test";

function buildRequirements(
  overrides?: Partial<{
    asset: string;
    amount: string;
    payTo: string;
    extra: Record<string, unknown>;
  }>,
): Parameters<ExactCasperScheme["enhancePaymentRequirements"]>[0] {
  return {
    scheme: "exact",
    network: testNetwork,
    asset: testAsset,
    amount: "1000000",
    payTo: testPayTo,
    maxTimeoutSeconds: 300,
    extra: {
      name: "TestToken",
      version: "1",
    },
    ...overrides,
  };
}

describe("ExactCasperScheme", () => {
  describe("parsePrice", () => {
    it("returns explicit AssetAmount with extra preserved", async () => {
      const scheme = new ExactCasperScheme();

      const result = await scheme.parsePrice(
        {
          amount: "1000000",
          asset: testAsset,
          extra: { name: "MyToken", version: "1" },
        },
        testNetwork,
      );

      expect(result.amount).toBe("1000000");
      expect(result.asset).toBe(testAsset);
      expect(result.extra).toMatchObject({ name: "MyToken", version: "1" });
    });

    it("returns explicit AssetAmount without extra with default empty extra", async () => {
      const scheme = new ExactCasperScheme();

      const result = await scheme.parsePrice(
        {
          amount: "500000",
          asset: testAsset,
        },
        testNetwork,
      );

      expect(result.amount).toBe("500000");
      expect(result.asset).toBe(testAsset);
      expect(result.extra).toEqual({});
    });

    it("throws for money string with no default asset or parser", async () => {
      const scheme = new ExactCasperScheme();

      await expect(scheme.parsePrice("1.00", testNetwork)).rejects.toThrow(
        "invalid_exact_casper_server_no_default_asset",
      );
    });

    it("throws for money float with no default asset or parser", async () => {
      const scheme = new ExactCasperScheme();

      await expect(scheme.parsePrice(1.0, testNetwork)).rejects.toThrow(
        "invalid_exact_casper_server_no_default_asset",
      );
    });

    it("returns custom parser result when registered", async () => {
      const scheme = new ExactCasperScheme();
      scheme.registerMoneyParser(async () => ({
        amount: "9999",
        asset: testAsset,
        extra: { name: "Custom", version: "2" },
      }));

      const result = await scheme.parsePrice(1.0, testNetwork);

      expect(result.amount).toBe("9999");
      expect(result.asset).toBe(testAsset);
      expect(result.extra).toMatchObject({ name: "Custom", version: "2" });
    });

    it("throws for invalid asset in map", async () => {
      const scheme = new ExactCasperScheme();

      await expect(
        scheme.parsePrice({ amount: "1000000", asset: "not-valid" }, testNetwork),
      ).rejects.toThrow("invalid_exact_casper_server_invalid_asset");
    });
  });

  describe("enhancePaymentRequirements", () => {
    it("returns requirements unchanged on happy path", async () => {
      const scheme = new ExactCasperScheme();

      const result = await scheme.enhancePaymentRequirements(
        buildRequirements({
          extra: { name: "TestToken", version: "1" },
        }),
        { x402Version: 1, scheme: "exact", network: testNetwork },
        [],
      );

      expect(result.amount).toBe("1000000");
      expect(result.asset).toBe(testAsset);
      expect(result.extra).toMatchObject({ name: "TestToken", version: "1" });
    });

    it("converts decimal amount using registered asset decimals", async () => {
      const scheme = new ExactCasperScheme();
      scheme.registerAsset(testNetwork, testAsset, 6);

      const result = await scheme.enhancePaymentRequirements(
        buildRequirements({ amount: "1.5" }),
        { x402Version: 1, scheme: "exact", network: testNetwork },
        [],
      );

      expect(result.amount).toBe("1500000");
    });

    it("throws when token name is missing", async () => {
      const scheme = new ExactCasperScheme();

      await expect(
        scheme.enhancePaymentRequirements(
          buildRequirements({ extra: { version: "1" } }),
          { x402Version: 1, scheme: "exact", network: testNetwork },
          [],
        ),
      ).rejects.toThrow("invalid_exact_casper_server_missing_token_name");
    });

    it("throws when token version is missing", async () => {
      const scheme = new ExactCasperScheme();

      await expect(
        scheme.enhancePaymentRequirements(
          buildRequirements({ extra: { name: "TestToken" } }),
          { x402Version: 1, scheme: "exact", network: testNetwork },
          [],
        ),
      ).rejects.toThrow("invalid_exact_casper_server_missing_token_version");
    });

    it("throws for invalid asset", async () => {
      const scheme = new ExactCasperScheme();

      await expect(
        scheme.enhancePaymentRequirements(
          buildRequirements({ asset: "too-short" }),
          { x402Version: 1, scheme: "exact", network: testNetwork },
          [],
        ),
      ).rejects.toThrow("invalid_exact_casper_server_invalid_asset");
    });

    it("throws for invalid payTo", async () => {
      const scheme = new ExactCasperScheme();

      await expect(
        scheme.enhancePaymentRequirements(
          buildRequirements({ payTo: "invalid" }),
          { x402Version: 1, scheme: "exact", network: testNetwork },
          [],
        ),
      ).rejects.toThrow("invalid_exact_casper_server_invalid_payto");
    });

    it("forwards extension keys from supportedKind extra", async () => {
      const scheme = new ExactCasperScheme();

      const result = await scheme.enhancePaymentRequirements(
        buildRequirements({ extra: { name: "T", version: "1" } }),
        {
          x402Version: 1,
          scheme: "exact",
          network: testNetwork,
          extra: { customKey: "customValue" },
        },
        ["customKey"],
      );

      expect(result.extra.customKey).toBe("customValue");
    });
  });

  describe("getAssetDecimals", () => {
    it("returns registered decimals", () => {
      const scheme = new ExactCasperScheme();
      scheme.registerAsset(testNetwork, testAsset, 9);

      expect(scheme.getAssetDecimals(testAsset, testNetwork)).toBe(9);
    });

    it("falls back to default of 9 for unknown assets", () => {
      const scheme = new ExactCasperScheme();

      expect(scheme.getAssetDecimals("unknown", testNetwork)).toBe(9);
    });
  });
});

File: go/x402/mechanisms/casper/exact/server/scheme_test.go
package server_test

import (
	"context"
	"testing"

	"casper_x402_facilitator/x402/mechanisms/casper"
	casperServer "casper_x402_facilitator/x402/mechanisms/casper/exact/server"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	x402 "github.com/x402-foundation/x402/go"
	"github.com/x402-foundation/x402/go/types"
)

const (
	testAsset   = "aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788"
	testNetwork = casper.NetworkCasperTestnet
)

func TestParsePrice_ExplicitAssetAmount(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	price := map[string]interface{}{
		"amount": "1000000",
		"asset":  testAsset,
		"extra": map[string]interface{}{
			"name":    "MyToken",
			"version": "1",
		},
	}

	result, err := s.ParsePrice(price, x402.Network(testNetwork))
	require.NoError(t, err)
	assert.Equal(t, "1000000", result.Amount)
	assert.Equal(t, testAsset, result.Asset)
	assert.Equal(t, "MyToken", result.Extra["name"])
}

func TestParsePrice_ExplicitNoExtra(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	price := map[string]interface{}{
		"amount": "500000",
		"asset":  testAsset,
	}

	result, err := s.ParsePrice(price, x402.Network(testNetwork))
	require.NoError(t, err)
	assert.Equal(t, "500000", result.Amount)
	assert.Equal(t, testAsset, result.Asset)
}

func TestParsePrice_MoneyString_NoDefaultAsset(t *testing.T) {
	s := casperServer.NewExactCasperScheme()

	_, err := s.ParsePrice("1.00", x402.Network(testNetwork))
	assert.ErrorContains(t, err, casperServer.ErrNoDefaultAsset)
}

func TestParsePrice_MoneyFloat_NoDefaultAsset(t *testing.T) {
	s := casperServer.NewExactCasperScheme()

	_, err := s.ParsePrice(1.0, x402.Network(testNetwork))
	assert.ErrorContains(t, err, casperServer.ErrNoDefaultAsset)
}

func TestParsePrice_WithCustomMoneyParser(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	s.RegisterMoneyParser(func(_ float64, _ x402.Network) (*x402.AssetAmount, error) {
		return &x402.AssetAmount{
			Amount: "9999",
			Asset:  testAsset,
			Extra:  map[string]interface{}{"name": "Custom", "version": "2"},
		}, nil
	})

	result, err := s.ParsePrice(1.0, x402.Network(testNetwork))
	require.NoError(t, err)
	assert.Equal(t, "9999", result.Amount)
}

func TestParsePrice_InvalidAssetInMap(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	price := map[string]interface{}{
		"amount": "1000000",
		"asset":  "not-valid",
	}

	_, err := s.ParsePrice(price, x402.Network(testNetwork))
	assert.ErrorContains(t, err, casperServer.ErrInvalidAsset)
}

func TestEnhancePaymentRequirements_HappyPath(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	req := types.PaymentRequirements{
		Scheme:  "exact",
		Network: testNetwork,
		Asset:   testAsset,
		Amount:  "1000000",
		PayTo:   "00aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788",
		Extra: map[string]interface{}{
			"name":    "TestToken",
			"version": "1",
		},
	}

	result, err := s.EnhancePaymentRequirements(context.Background(), req, types.SupportedKind{}, nil)
	require.NoError(t, err)
	assert.Equal(t, "1000000", result.Amount)
}

func TestEnhancePaymentRequirements_DecimalAmount(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	s.RegisterAsset(testNetwork, testAsset, 6)
	req := types.PaymentRequirements{
		Scheme:  "exact",
		Network: testNetwork,
		Asset:   testAsset,
		Amount:  "1.5",
		PayTo:   "00aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788",
		Extra: map[string]interface{}{
			"name":    "TestToken",
			"version": "1",
		},
	}

	result, err := s.EnhancePaymentRequirements(context.Background(), req, types.SupportedKind{}, nil)
	require.NoError(t, err)
	assert.Equal(t, "1500000", result.Amount)
}

func TestEnhancePaymentRequirements_MissingTokenName(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	req := types.PaymentRequirements{
		Scheme:  "exact",
		Network: testNetwork,
		Asset:   testAsset,
		Amount:  "1000000",
		PayTo:   "00aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788",
		Extra:   map[string]interface{}{"version": "1"},
	}

	_, err := s.EnhancePaymentRequirements(context.Background(), req, types.SupportedKind{}, nil)
	assert.ErrorContains(t, err, casperServer.ErrMissingTokenName)
}

func TestEnhancePaymentRequirements_InvalidAsset(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	req := types.PaymentRequirements{
		Scheme:  "exact",
		Network: testNetwork,
		Asset:   "too-short",
		Amount:  "1000000",
		PayTo:   "00aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788",
		Extra:   map[string]interface{}{"name": "T", "version": "1"},
	}

	_, err := s.EnhancePaymentRequirements(context.Background(), req, types.SupportedKind{}, nil)
	assert.ErrorContains(t, err, casperServer.ErrInvalidAsset)
}

func TestEnhancePaymentRequirements_InvalidPayTo(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	req := types.PaymentRequirements{
		Scheme:  "exact",
		Network: testNetwork,
		Asset:   testAsset,
		Amount:  "1000000",
		PayTo:   "invalid",
		Extra:   map[string]interface{}{"name": "T", "version": "1"},
	}

	_, err := s.EnhancePaymentRequirements(context.Background(), req, types.SupportedKind{}, nil)
	assert.ErrorContains(t, err, casperServer.ErrInvalidPayTo)
}

func TestEnhancePaymentRequirements_ExtensionKeysForwarded(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	req := types.PaymentRequirements{
		Scheme:  "exact",
		Network: testNetwork,
		Asset:   testAsset,
		Amount:  "1000000",
		PayTo:   "00aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788",
		Extra:   map[string]interface{}{"name": "T", "version": "1"},
	}
	sk := types.SupportedKind{Extra: map[string]interface{}{"customKey": "customValue"}}

	result, err := s.EnhancePaymentRequirements(context.Background(), req, sk, []string{"customKey"})
	require.NoError(t, err)
	assert.Equal(t, "customValue", result.Extra["customKey"])
}

func TestGetAssetDecimals_Registered(t *testing.T) {
	s := casperServer.NewExactCasperScheme()
	s.RegisterAsset(testNetwork, testAsset, 9)

	d := s.GetAssetDecimals(testAsset, x402.Network(testNetwork))
	assert.Equal(t, 9, d)
}

func TestGetAssetDecimals_DefaultFallback(t *testing.T) {
	s := casperServer.NewExactCasperScheme()

	d := s.GetAssetDecimals("unknown", x402.Network(testNetwork))
	assert.Equal(t, 9, d)
}
